### PR TITLE
[R-package] De-couple evaluation monitoring from booster verbosity

### LINKED ISSUE #6162: [R-package] `lgb.cv` mixes booster verbosity with evaluation metric monitoring
When one calls `lgb.cv`, one usually wants to monitor the evaluation metric in the training and validation set as the fitting procedure happens, and one usually doesn't want to know about every internal detail of what's happening inside each boosting round.

`lgb.cv` takes a `verbosity` parameter, with levels for fatal/warning/info/debug, but the messages about the metric of interest in the training and validation data do not correspond to any of those. Currently, they'd be bumped together with `info`, even though they don't say it in the message.

But there's no current arrangement in which one would only get the messages about the metrics and not the internal messages from each boosting round - example:
```r
library(lightgbm)
data(mtcars)
y <- mtcars$mpg
x <- mtcars[, -1] |> as.matrix()
result <- lgb.cv(
    data = lgb.Dataset(x, label=y),
    params = list(
        objective = "regression",
        metric = "l2",
        min_data_in_leaf = 5
    ),
    nrounds = 5,
    nfold = 3,
    verbose = 1,
    eval_train_metric = TRUE
)
```
```
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000055 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 74
[LightGBM] [Info] Number of data points in the train set: 22, number of used features: 10
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000044 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 74
[LightGBM] [Info] Number of data points in the train set: 21, number of used features: 10
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000041 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 74
[LightGBM] [Info] Number of data points in the train set: 21, number of used features: 10
[LightGBM] [Info] Start training from score 20.604546
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Start training from score 19.761905
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Start training from score 19.880952
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1] "[1]:  train's l2:29.7871+2.01926  valid's l2:31.7638+4.84005"
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1] "[2]:  train's l2:25.4878+1.93599  valid's l2:28.1109+4.93162"
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1] "[3]:  train's l2:21.9533+1.92091  valid's l2:25.3463+5.08398"
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1] "[4]:  train's l2:18.9155+1.72977  valid's l2:23.1842+5.24103"
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1] "[5]:  train's l2:16.4374+1.55234  valid's l2:21.0872+5.31824"
```

Whereas one would usually want to see only these:
```
[1] "[1]:  train's l2:29.7871+2.01926  valid's l2:31.7638+4.84005"
[1] "[2]:  train's l2:25.4878+1.93599  valid's l2:28.1109+4.93162"
[1] "[3]:  train's l2:21.9533+1.92091  valid's l2:25.3463+5.08398"
[1] "[4]:  train's l2:18.9155+1.72977  valid's l2:23.1842+5.24103"
[1] "[5]:  train's l2:16.4374+1.55234  valid's l2:21.0872+5.31824"
```

And without the `[1]` that gets added by R's print function.

Should ideally have a separate parameter for deciding whether to print the metric that wouldn't be tied to the booster's internals.

### PR DESCRIPTION
**Pull request by david-cortes**
Fixes https://github.com/microsoft/LightGBM/issues/6162

This PR decouples the verbosity of boosters from the printing of evaluation metrics during cross-validation when calling `lgb.cv`.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for bringing this up, I support separating configuration of when these logs get printed.

But instead of introducing a new keyword argument in this interface, I think it'd be better to handle "print eval metrics every `eval_freq` iterations" via callbacks.

The same thing was done in the Python package here... keyword argument `verbose_eval`, whose purpose was very similar to the `print_metrics` proposed in this PR, was deprecated in #4574 and removed in #4878, in favor of users passing the `lgb.log_evaluation()` callback.

XGBoost does something like that `xgb.cv()` in its R package... exporting a callback `cb.print.evaluation()` ...

https://github.com/dmlc/xgboost/blob/6c0a190f6d12d2ba6a1cabd7741881ea1913d433/R-package/NAMESPACE#L23

... and also optionally adding it if it wasn't provided but a keyword argument `verbose` is passed

https://github.com/dmlc/xgboost/blob/6c0a190f6d12d2ba6a1cabd7741881ea1913d433/R-package/R/xgb.cv.R#L164-L166.

So rather than introduce a new R-package-only parameter that uses and maintainers have to think about, I think it'd be preferable to:

1. export callback `cb_print_evaluation()` (changing keyword argument `period` to be named `eval_freq` for consistency)

https://github.com/microsoft/LightGBM/blob/16004228e75dce0a577979875506eec772dc7ba5/R-package/R/callback.R#L74

2. make `callbacks = list(cb_print_evaluation(1L))` the default in `lgb.cv()`
3. remove any logic in `lgb.cv()` that conditionally adds that to the list of callbacks

What do you think? also tagging @mayer79 @jmoralez 

If you support that path but aren't interested in doing the work of exporting that callback, let me know and I'll happily do it.

---

tagging related issue: #2479

**Comment by david-cortes** [comment]
Regarding the exporting of callbacks, I think a first step would be to document their structure and what they are passed. Currently, it's very unclear to me what exactly should a callback conform to and what exactly does the variable they get passed have inside at each point. I'll leave the exporting and documentation to package maintainers here.

Regarding the proposed logic - I am not sure if that's the best way. Currently, there are two separate callbacks for 'record' and for 'eval_freq' which feel like they should be doing the same calculation, but with the printing one potentially emitting logs less frequently. I think it'd be easier to understand if there was only one such callback with a printing frequency argument and perhaps a flag for signaling an early stop.

**Comment by jameslamb** [comment]
> Regarding the exporting of callbacks, I think a first step would be to document their structure and what they are passed.

I agree! I can work on that (tracked in #2479).

> I think it'd be easier to understand if there was only one such callback with a printing frequency argument and perhaps a flag for signaling an early stop.

That's great feedback, thank you.

Thankfully these aren't yet really part of the R package's public API so we have time to consider alternative designs.
